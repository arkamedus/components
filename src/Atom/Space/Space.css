.space {
    display: flex;
    flex-wrap: wrap;
}

.space.gap {
    column-gap: var(--leading);
    row-gap: var(--leading);
}



/* Direction */
.direction-horizontal {
    flex-direction: row;
}

.direction-vertical {
    flex-direction: column;
}

/* Alignment */
.align-default {
    align-items: flex-start;
}

.align-center {
    align-items: center;
}

.align-start {
    align-items: flex-start;
}

.align-end {
    align-items: flex-end;
}

.align-normal {
    align-items: normal;
}

/* Justification */
.justify-default {
    justify-content: normal;
}

.justify-center {
    justify-content: center;
}

.justify-start {
    justify-content: flex-start;
}

.justify-end {
    justify-content: flex-end;
}

.justify-stretch {
    justify-content: stretch;
}

.justify-stretch > * {
    flex-grow: 1;
}

.justify-around {
    justify-content: space-around;
}

.justify-between {
    justify-content: space-between;
}

.justify-evenly {
    justify-content: space-evenly;
}
