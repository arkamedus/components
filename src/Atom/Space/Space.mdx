import { Controls, Meta, Primary, Stories } from "@storybook/blocks";
import meta from './Space.stories';

<Meta of={meta} />

# Space

The `Space` component provides a flexible container to arrange its children with configurable spacing, alignment, and justification. It allows you to control the gap between items, set the layout direction (default, horizontal, or vertical), and align or distribute items along both the cross and main axes.

<Primary />
<Controls />

---

# Space Component

The `Space` component is designed to simplify layout creation by managing spacing and alignment for its children. It accepts several key props:

- **gap**: Toggles the gap between items.
- **direction**: Determines the layout orientation (options: `default`, `horizontal`, `vertical`).
- **align**: Controls the alignment of items along the cross axis (options: `default`, `normal`, `center`, `start`, `end`).
- **justify**: Sets the distribution of items along the main axis (options: `default`, `center`, `start`, `end`, `stretch`, `around`, `between`, `evenly`).

This component is ideal for creating consistent, responsive layouts without the need to write additional spacing or alignment CSS.

<Stories />
