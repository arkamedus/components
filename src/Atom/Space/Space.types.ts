import {
	ComponentAlignType,
	ComponentJustifyType,
	CoreComponentLayoutDirectionType,
	CoreLayoutProps,
} from "../../Core/Core.types";

export interface SpaceProps extends CoreLayoutProps {
	direction?: CoreComponentLayoutDirectionType;
	align?: ComponentAlignType;
	justify?: ComponentJustifyType;
	wide?: boolean;
}
