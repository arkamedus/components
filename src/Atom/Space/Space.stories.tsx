import React from "react";
import {Meta, StoryFn, StoryObj} from "@storybook/react";
import Space from "./Space";
import Paragraph from "../Paragraph/Paragraph";
import Button from "../Button/Button";
import DebugLayer from "../DebugLayer/DebugLayer";

const meta:Meta<typeof Space> =  {
    title: "Design System/Atomic/Space",
    component: Space,
    argTypes: {
        gap: {
            control: "boolean",
            description: "Toggle gap between items",
        },
        direction: {
            control: { type: "select" },
            options: ["default", "horizontal", "vertical"],
            description: "Layout direction of the space",
        },
        align: {
            control: { type: "select" },
            options: ["default", "normal", "center", "start", "end"],
            description: "Alignment of items along the cross axis",
        },
        justify: {
            control: { type: "select" },
            options: [
                "default",
                "center",
                "start",
                "end",
                "stretch",
                "around",
                "between",
                "evenly",
            ],
            description: "Justification of items along the main axis",
        },
        onClick: { action: "clicked", table: { disable: true } }, // Example action (if needed)
    },
    args: {
        gap: false,
        direction: "default",
        align: "default",
        justify: "default",
    },
} as Meta<typeof Space>;
export default meta;
type Story = StoryObj<typeof Button>;


const card = (
    <div style={{ border: "1px solid #999", padding: "10px" }}>
        <h1>Card Title</h1>
        <Paragraph>Card paragraph</Paragraph>
    </div>
);

const Template: StoryFn<typeof Space> = (args) => (
    <Space {...args}>
        <DebugLayer label="DebugLayer" />
        {card}
    </Space>
);

export const Default = Template.bind({});
Default.args = {
    gap: false,
    direction: "default",
    align: "default",
    justify: "default",
};

export const Gap = Template.bind({});
Gap.args = {
    gap: true,
};

export const Vertical = Template.bind({});
Vertical.args = {
    gap: true,
    direction: "vertical",
};

export const Center = Template.bind({});
Center.args = {
    align: "center",
    justify: "center",
};

export const SpaceBetween = () => (
    <Space justify="between">
        <DebugLayer label="DebugLayer" />
        <DebugLayer label="DebugLayer" />
    </Space>
);

export const StretchGap = () => (
    <Space justify="stretch" gap>
        <DebugLayer label="DebugLayer" />
        <DebugLayer label="DebugLayer">
            <Paragraph>
                Content Here
                <br />
                To make it taller.
            </Paragraph>
        </DebugLayer>
    </Space>
);
