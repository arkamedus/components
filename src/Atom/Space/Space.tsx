import React from "react";

import { SpaceProps } from "./Space.types";

import "./Space.css";

const Space: React.FC<SpaceProps> = ({
	id, key,
	style,
	children,
	className,
	gap,
	direction,
	align,
	justify,
	wide,
	noWrap,
	onClick
}) => {
	let classNames = ["oakd", "space"];
	if (className) classNames.push(className);
	if (direction) classNames.push(`direction-${direction}`);
	if (align) classNames.push(`align-${align}`);
	if (justify) classNames.push(`justify-${justify}`);
	if (noWrap) classNames.push(`nowrap`);
	if (gap) {
		classNames.push("gap");
	}
	if (wide) {
		classNames.push("wide");
	}

	const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
		if (onClick) onClick(event);
	};

	return (
		<span id={id} key={key} onClick={handleClick} data-testid="Space" style={style} className={classNames.join(" ")}>
			{children}
		</span>
	);
};

export default Space;
