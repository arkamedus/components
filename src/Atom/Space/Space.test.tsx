import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import Space from "./Space";

describe("Space Component", () => {
  const renderComponent = (props = {}) =>
    render(
      <Space data-testid="Space" {...props}>
        <div>Item 1</div>
        <div>Item 2</div>
      </Space>,
    );

  it("should render without errors", () => {
    const { getByTestId } = renderComponent();
    expect(getByTestId("Space")).toBeInTheDocument();
  });

  it("should apply default classes", () => {
    const { getByTestId } = renderComponent();
    expect(getByTestId("Space")).toHaveClass("space");
  });

  it("should render children correctly", () => {
    const { getByTestId } = renderComponent();
    expect(getByTestId("Space")).toHaveTextContent("Item 1");
    expect(getByTestId("Space")).toHaveTextContent("Item 2");
  });

  it("should apply gap class when gap prop is true", () => {
    const { getByTestId } = renderComponent({ gap: true });
    expect(getByTestId("Space")).toHaveClass("gap");
  });

  it("should apply direction class correctly", () => {
    const { getByTestId } = renderComponent({ direction: "vertical" });
    expect(getByTestId("Space")).toHaveClass("direction-vertical");
  });

  it("should apply alignment class correctly", () => {
    const { getByTestId } = renderComponent({ align: "center" });
    expect(getByTestId("Space")).toHaveClass("align-center");
  });

  it("should apply justification class correctly", () => {
    const { getByTestId } = renderComponent({ justify: "between" });
    expect(getByTestId("Space")).toHaveClass("justify-between");
  });

  it("should match snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
