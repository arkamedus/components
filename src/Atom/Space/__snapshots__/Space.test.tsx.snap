// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Space Component should match snapshot 1`] = `
<div>
  <div
    class="oakd space"
    data-testid="Space"
  >
    <div>
      Item 1
    </div>
    <div>
      Item 2
    </div>
  </div>
</div>
`;
